sygnal/database.py

```diff
@@ -0,0 +1,77 @@
+# -*- coding: utf-8 -*-
+# Copyright 2014 matrix.org
+# Copyright 2019 The Matrix.org Foundation C.I.C.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import sqlite3
+from concurrent.futures import ThreadPoolExecutor
+
+from twisted.internet.defer import Deferred
+
+logger = logging.getLogger(__name__)
+
+
+class Database:
+    def __init__(self, dbfile, twisted_reactor):
+        self.twisted_reactor = twisted_reactor
+        self.dbfile = dbfile
+        # SQLite is blocking, so run queries in a separate thread
+        self.dbexecutor = ThreadPoolExecutor(max_workers=1)
```

Comment:
Might be worth looking at `aiosqlite`?

Reply:
AFAICT `aiosqlite` does the same thing (run all queries in serial on a worker thread) so I'm not sure if it would gain anything.

Reply:
Sure, but it saves me from having to figure out if your implementation is doing the right thing :)

Reply:
On another note, it would break testing since the test environment cannot use `asyncio` for async; pure Twisted only.

Reply:
Ah, then you probably want https://twistedmatrix.com/documents/current/core/howto/rdbms.html.